// nodemailer-mailgun-transport/es5/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
var mailgun = require("mailgun-js");
var Attachment = require("mailgun-js/lib/attachment");
var consolidate = require("consolidate");
var packageData = require("../package.json");
var whitelist = [
    ["replyTo", "h:Reply-To"],
    [/^h:/],
    [/^v:/],
    ["from"],
    ["to"],
    ["cc"],
    ["bcc"],
    ["subject"],
    ["text"],
    ["template"],
    ["html"],
    ["attachment"],
    ["inline"],
    ["recipient-variables"],
    ["o:tag"],
    ["o:campaign"],
    ["o:dkim"],
    ["o:deliverytime"],
    ["o:testmode"],
    ["o:tracking"],
    ["o:tracking-clicks"],
    ["o:tracking-opens"],
    ["o:require-tls"],
    ["o:skip-verification"],
    ["X-Mailgun-Variables"]
];
var applyKeyWhitelist = function (mail) {
    return Object.keys(mail).reduce(function (acc, key) {
        var _a;
        var targetKey = whitelist.reduce(function (result, _a) {
            var cond = _a[0], target = _a[1];
            if (result) {
                return result;
            }
            if ((cond.exec && cond.exec(key)) || cond === key) {
                return target || key;
            }
            return null;
        }, null);
        if (!targetKey || !mail[key]) {
            return acc;
        }
        return __assign(__assign({}, acc), (_a = {}, _a[targetKey] = mail[key], _a));
    }, {});
};
var renderTemplate = function (template) { return __awaiter(_this, void 0, void 0, function () {
    var engine, name, _a, context, html;
    return __generator(this, function (_b) {
        switch (_b.label) {
            case 0:
                if (!template || !template.name || !template.engine) {
                    return [2 /*return*/, {}];
                }
                engine = template.engine, name = template.name, _a = template.context, context = _a === void 0 ? {} : _a;
                return [4 /*yield*/, consolidate[engine](name, context)];
            case 1:
                html = _b.sent();
                return [2 /*return*/, { template: null, html: html }];
        }
    });
}); };
var makeMailgunAttachments = function (attachments) {
    if (attachments === void 0) { attachments = []; }
    var _a = attachments.reduce(function (results, item) {
        var data = typeof item.content === "string"
            ? Buffer.from(item.content, item.encoding)
            : item.content || item.path || undefined;
        // mailgunjs does not encode content string to a buffer
        var attachment = new Attachment({
            data: data,
            filename: item.cid || item.filename || undefined,
            contentType: item.contentType || undefined,
            knownLength: item.knownLength || undefined
        });
        var attachmentAttachments = results[0], inlineAttachments = results[1];
        return [
            attachmentAttachments.concat(!item.cid ? attachment : []),
            inlineAttachments.concat(item.cid ? attachment : [])
        ];
    }, [[], []]), attachment = _a[0], inline = _a[1];
    return __assign(__assign({}, (attachment.length ? { attachment: attachment } : {})), (inline.length ? { inline: inline } : {}));
};
var makeAllTextAddresses = function (mail) {
    var keys = ["from", "to", "cc", "bcc", "replyTo"];
    var makeTextAddresses = function (addresses) {
        var validAddresses = [].concat(addresses).filter(Boolean);
        var textAddresses = validAddresses.map(function (item) {
            return item.address
                ? item.name
                    ? item.name + " <" + item.address + ">"
                    : item.address
                : typeof item === "string"
                    ? item
                    : null;
        });
        return textAddresses.filter(Boolean).join();
    };
    var result = keys.reduce(function (result, key) {
        var _a;
        var textAddresses = makeTextAddresses(mail[key]);
        if (!textAddresses) {
            return result;
        }
        return __assign(__assign({}, result), (_a = {}, _a[key] = textAddresses, _a));
    }, {});
    return result;
};
var send = function (mailgunSend) { return function (_a, callback) {
    var mail = _a.data;
    return __awaiter(_this, void 0, void 0, function () {
        var addresses, attachments, template, extendedMail, whitelistedMail, result, error_1;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 3, , 4]);
                    addresses = makeAllTextAddresses(mail);
                    attachments = makeMailgunAttachments(mail.attachments);
                    return [4 /*yield*/, renderTemplate(mail.template)];
                case 1:
                    template = _b.sent();
                    extendedMail = __assign(__assign(__assign(__assign({}, mail), addresses), attachments), template);
                    whitelistedMail = applyKeyWhitelist(extendedMail);
                    return [4 /*yield*/, mailgunSend(whitelistedMail)];
                case 2:
                    result = _b.sent();
                    callback(null, __assign(__assign({}, result), { messageId: result.id }));
                    return [3 /*break*/, 4];
                case 3:
                    error_1 = _b.sent();
                    callback(error_1);
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    });
}; };
var transport = function (options) {
    if (options === void 0) { options = {}; }
    var messages = mailgun({
        apiKey: options.auth.api_key,
        domain: options.auth.domain || "",
        proxy: options.proxy || false,
        host: options.host || "api.mailgun.net",
        protocol: options.protocol || "https:",
        port: options.port || 443
    }).messages();
    var mailgunSend = function (mail) { return messages.send(mail); };
    return {
        name: "Mailgun",
        version: packageData.version,
        send: send(mailgunSend),
        messages: messages,
        options: options
    };
};
transport._send = send;
transport._makeAllTextAddresses = makeAllTextAddresses;
transport._makeMailgunAttachments = makeMailgunAttachments;
transport._renderTemplate = renderTemplate;
transport._applyKeyWhitelist = applyKeyWhitelist;
module.exports = transport;
